penrose-web/src/compiler/Style.ts

```diff
@@ -1,7 +1,2068 @@
+  return Object.entries(selEnv.sTypeVarMap)
+    .reduce((acc, curr) => mergeMapping(selEnv.varProgTypeMap, acc, curr), varEnv);
+};
+
+// ported from `checkPair`, `checkSel`, and `checkNamespace`
+const checkHeader = (varEnv: Env, header: Header): SelEnv => {
+  if (header.tag === "Selector") {
+    // Judgment 7. G |- Sel ok ~> g
+    const sel: Selector = header;
+    const selEnv_afterHead = checkDeclPatternsAndMakeEnv(varEnv, initSelEnv(), sel.head.contents);
+    // Check `with` statements
+    // TODO: Did we get rid of `with` statements?
+    const selEnv_decls = checkDeclPatternsAndMakeEnv(varEnv, selEnv_afterHead, safeContentsList(sel.with));
+
+    const relErrs = checkRelPatterns(mergeEnv(varEnv, selEnv_decls), safeContentsList(sel.where));
+
+    // TODO(error): The errors returned in the top 3 statements
+    return {
+      ...selEnv_decls,
+      sErrors: selEnv_decls.sErrors.concat(relErrs) // COMBAK: Reverse the error order?
+    };
+  } else if (header.tag === "Namespace") {
+    // TODO(error)
+    return initSelEnv();
+  } else throw Error("unknown Style header tag");
+}
+
+// Returns a sel env for each selector in the Style program, in the same order
+// previously named `checkSels`
+export const checkSelsAndMakeEnv = (varEnv: Env, prog: HeaderBlock[]): SelEnv[] => {
+  const selEnvs: SelEnv[] = prog.map(e => {
+    const res = checkHeader(varEnv, e.header);
+    // Put selector AST in just for debugging
+    res.header = { tag: "Just", contents: e.header };
+    return res;
+  });
+  // const errors = ... TODO(errors)
+  return selEnvs; // TODO
+};
+
+//#endregion
+
+//#region Types and code for finding substitutions
+
+// Judgment 20. A substitution for a selector is only correct if it gives exactly one
+//   mapping for each Style variable in the selector. (Has test)
+export const fullSubst = (selEnv: SelEnv, subst: Subst): boolean => {
+  // Check if a variable is a style variable, not a substance one
+  const isStyVar = (e: string): boolean => selEnv.varProgTypeMap[e][0].tag === "StyProgT";
+
+  // Equal up to permutation (M.keys ensures that there are no dups)
+  const selStyVars = Object.keys(selEnv.sTypeVarMap).filter(isStyVar);
+  const substStyVars = Object.keys(subst);
+  // Equal up to permutation (keys of an object in js ensures that there are no dups)
+  return _.isEqual(selStyVars.sort(), substStyVars.sort());
+};
+
+// Check that there are no duplicate keys or vals in the substitution (Has test)
+export const uniqueKeysAndVals = (subst: Subst): boolean => {
+  // All keys already need to be unique in js, so only checking values
+  const vals = Object.values(subst);
+  const valsSet = {};
+
+  for (let i = 0; i < vals.length; i++) {
+    valsSet[vals[i]] = 0; // This 0 means nothing, we just want to create a set of values
+  }
+
+  // All entries were unique if length didn't change (ie the nub didn't change)
+  return Object.keys(valsSet).length === vals.length;
```

Comment:
Not sure what this function is doing, just looking for unique keys? If so, use `uniq` (https://lodash.com/docs/4.17.15#uniq)?